src/Google.Api.Gax/Platform.cs

```diff
@@ -0,0 +1,150 @@
+            ZoneName = json["instance"]["zone"].ToString();
+        }
+
+        public string MetadataJson { get; }
+        public string ProjectId { get; }
+        public string InstanceId { get; }
+        public string ZoneName { get; }
+    }
+
+    public sealed class GaePlatformDetails
+    {
+        public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeVersion)
+        {
+            ProjectId = gcloudProject;
+
+        }
+
+        public string ProjectId { get; }
+        public string ModuleId { get; }
+        public string VersionId { get; }
+    }
+
+    /// <summary>
+    /// Information about the current execution platform.
+    /// Supported execption platforms are Google App Engine (GAE) and Google Compute Engine (GCE).
+    /// </summary>
+    public sealed class Platform
+    {
+        private readonly static Lazy<Task<Platform>> s_instance = new Lazy<Task<Platform>>(LoadInstance);
+
+        /// <summary>
+        /// Asyncrhonously get execution platform information.
+        /// </summary>
+        /// <returns>A task containing the execution platform information.</returns>
+        public static Task<Platform> InstanceAsync() => s_instance.Value;
+
+        /// <summary>
+        /// Get execution platform information. This may block for up to one second.
+        /// </summary>
+        /// <returns>Execution platform information.</returns>
+        public static Platform Instance() => InstanceAsync().Result;
+
+        private static async Task<GcePlatformDetails> LoadGceDetails()
+        {
+            const string metadataUrl = "http://metadata.google.internal/computeMetadata/v1?recursive=true";
+            const string metadataFlavorKey = "Metadata-Flavor";
+            const string metadataFlavorValue = "Google";
+            const long maxContentLength = 512 * 1024; // Maximum allowed metadata size.
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, metadataUrl);
+            httpRequest.Headers.Add(metadataFlavorKey, metadataFlavorValue); // Required for any query.
+            var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(1000)); // 1000 found to be reasonable.
+            var httpClient = new HttpClient();
+            var response = await httpClient.SendAsync(httpRequest, cts.Token);
+            IEnumerable<string> metadataValues;
+            if (response.StatusCode == HttpStatusCode.OK
+                && response.Content.Headers.TryGetValues(metadataFlavorKey, out metadataValues)
+                && metadataValues.Contains(metadataFlavorValue)
+                && response.Content.Headers.ContentLength < maxContentLength)
+            {
+                // Valid response from metadata server.
+                string metadataJson = await response.Content.ReadAsStringAsync();
+                try
+                {
+                    return new GcePlatformDetails(metadataJson);
+                }
+                catch
```

Comment:
True, but I'm not sure what to do here.
I've added a TODO to decide what to do.